patroni/dcs/__init__.py

```diff
@@ -402,21 +533,37 @@ def from_node(version: _Version, value: Union[str, Dict[str, str]]) -> 'Failover
         return Failover(version, data.get('leader'), data.get('member'), data.get('scheduled_at'))
 
     def __len__(self) -> int:
+        """Implement ``len`` function capability."""
```

Comment:
I think it would be worth to mention that this method is called when something like `bool(cluster.failover)` or `if cluster.failover` is executed.

The reason it was done this way because Patroni need to create/update/delete failover key atomically, but ZooKeeper implements atomic operations it very weirdly. In ZooKeeper these operations are based on version of a given ZNode, which is reset to 0 every time when ZNode is dropper and created. That is, we have a /failover key with version=0 and want to remove it, but meanwhile some other thread removed and added it back. As a result we will be removing not a key we wanted to remove.

To mitigate this behavior Patroni just empties the /failover key instead of removing it. When we read the cluster state from DCS, the Failover object will be created, because we need to know its version for the next update.

Without `__len__` we would have to write something like `if cluster.failover and (cluster.failover.leader or cluster.failover.candidate)` most of the time, and now we can just do `if cluster.failover`, which will be evaluated to `False` if object is "empty".

Reply:
Thanks that makes sense!

Reply:
See suggestion https://github.com/zalando/patroni/pull/2777/files#r1279176340